Спринты 10 и 11

ihatelags opened the pull request:
feat: перенес хранение в DB
feat: добавил эндпоинты /genres и /mpa

ihatelags committed:
Спринты 10 и 11
feat: перенес хранение в DB
feat: добавил эндпоинты /genres и /mpa

ihatelags committed:
удалил старые тесты

ihatelags committed:
фиксы после тестов в пулл-реквесте

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,205 @@
+
+@Component
+@Primary
+public class DbFilmStorage implements FilmStorage{
```
пробел перед { :)))

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,205 @@
+    public Film createFilm(Film film) {
+        String sql = "INSERT INTO films (film_name, description, release_date, duration, rate, rating_id)" +
+                "VALUES (?, ?, ?, ?, ?, ?);" ;
+        SqlRowSet sqlRows = jdbcTemplate.queryForRowSet("SELECT COUNT(film_id) AS count FROM films ;");
```
здесь лучше воспользоваться кейхолдером:
`KeyHolder keyHolder = new GeneratedKeyHolder();`
не всегда ид идут по порядку, можно счетчиком просчитаться)

ihatelags replied:
Код с keyholder и prepareStatement как-то сложно читается. Нашел в гугле как сделать через SimpleJdbcInsert и executeAndReturnKey.

utausheva replied:
да, отличный подход!

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,205 @@
+    }
+
+    @Override
+    public void addLike(Long filmId, Long userId) {
```
методы для лайков лучше перенести в дао лайков

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,205 @@
+        return genres;
+    }
+
+    public Genre getGenre(long id) {
```
2 метода для жанров - в дао жанров (они ведь никак не относятся к фильмам, это самостоятельные справочники)

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFilmStorage.java:
```diff
@@ -0,0 +1,205 @@
+        );
+    }
+
+    public MpaRating getMpa(long id) {
```
для мпа тоже перенести методы в дао мпа

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,44 @@
+public class GenreController {
+    private static final Logger log = LoggerFactory.getLogger(GenreController.class);
+
+    private DbFilmStorage filmStorage;
```
здесь будет использоваться хранилище жанров (либо сервис жанров, но можно его не делать, т.к. жанры - это справочник)

utausheva replied:
но обязательно не класс, а интерфейс

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/controller/GenreController.java:
```diff
@@ -0,0 +1,44 @@
+
+    @GetMapping("/genres/{genreId}")
+    public Genre getGenreById(@PathVariable long genreId) {
+        if (genreId > 6 || genreId < 1) {
```
проверку на больше 6 лучше убрать, т.к. при добавлении нового жанра придется переделывать и контроллер, и хардкодить ограничения нехорошо

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/controller/MpaController.java:
```diff
@@ -0,0 +1,52 @@
+     */
+    @GetMapping("/mpa/{id}")
+    public MpaRating getMpa(@PathVariable long id) {
+        if (id > 5 || id < 1) {
```
то же, что для жанров

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/controller/MpaController.java:
```diff
@@ -0,0 +1,52 @@
+public class MpaController {
+    private static final Logger log = LoggerFactory.getLogger(MpaController.class);
+
+    private DbFilmStorage filmStorage;
```
здесь будет хранилище мпа (или сервис мпа)

utausheva reviewed (commented):
Привет!
В целом отлично, понравились запросы в БД, очень четкие, легко читаются, оптимизированы)
Но надо поработать над моделью: не хватает дао для жанров и мпа, также неплохо было бы добавить дао лайков и друзей (не забудь про интерфейсы), чтобы код был логически разделен на разные классы.

ihatelags committed:
фиксы после ревью:
- вынес методы в отдельные классы (жанры - GenreStorage и тд)
- заменил способ генерации ид
- убрал лишние проверки в контроллерах

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFriendStorage.java:
```diff
@@ -0,0 +1,93 @@
+public class DbFriendStorage implements FriendStorage {
+
+    private final JdbcTemplate jdbcTemplate;
+    private final UserStorage userStorage;
```
хранилища не должны зависеть друг от друга

utausheva commented on src/main/java/ru/yandex/practicum/filmorate/storage/DbFriendStorage.java:
```diff
@@ -0,0 +1,93 @@
+
+    @Override
+    public List<User> getFriends(Long userId) {
+        String sql = "SELECT * FROM friendships WHERE user_id = ? ;";
```
тут можно вот такой запрос:
"select * from USERS, FRIENDS where USERS.USER_ID = FRIENDS.FRIEND_ID AND FRIENDS.USER_ID = ?"
только названия таблиц и столбцов подставь свои

utausheva reviewed (approved):
В целом ок, если есть желание, то подправь одну зависимость)
Можно мерджить)
